fabtests/common/shared.c

```diff
@@ -3121,18 +3127,36 @@ int ft_parse_rma_opts(int op, char *optarg, struct fi_info *hints,
 	return 0;
 }
 
-void ft_fill_buf(void *buf, size_t size)
+int ft_fill_buf(void *buf, size_t size)
 {
 	char *msg_buf;
 	int msg_index = 0;
 	size_t i;
 
-	msg_buf = (char *) buf;
+	if (opts.iface != FI_HMEM_SYSTEM) {
+		msg_buf = malloc(size);
+		if (!msg_buf) {
+			free(msg_buf);
+			return -FI_EIO;
+		}
+	} else {
+		msg_buf = (char *) buf;
+	}
+
 	for (i = 0; i < size; i++) {
 		msg_buf[i] = integ_alphabet[msg_index];
 		if (++msg_index >= integ_alphabet_length)
 			msg_index = 0;
 	}
+
+	if (opts.iface != FI_HMEM_SYSTEM) {
+		if (ft_hmem_copy_to(opts.iface, opts.device, buf, msg_buf, size)) {
+			free(msg_buf);
+			return -FI_EIO;
+		}
+	}
```

Comment:
msg_buf is leaked  here.  Always need to free at the end, even if ft_hmem_copy_to() succeeds.

Restructure the code to have a common exit where msg_buf is freed.  For example:
```
    int ret = 0;
    ....
        ret = ft_hmem_copy_to(...)
        if (ret)
            goto out;

    ...

out:
    if (opts.iface != FI_HMEM_SYSTEM)
        free(msg_buf)
    return ret;
```